processor/ratelimitprocessor/processor.go

```diff
@@ -135,35 +172,105 @@ func (r *ProfilesRateLimiterProcessor) Capabilities() consumer.Capabilities {
 	return consumer.Capabilities{MutatesData: false}
 }
 
+func getTelemetryAttrs(ctx context.Context, metadataKeys []string, err error) []attribute.KeyValue {
+	attrs := getAttrsFromContext(ctx, metadataKeys)
+
+	switch {
+	case err == nil:
+		attrs = append(attrs,
+			telemetry.WithReason(telemetry.StatusUnderLimit),
+			telemetry.WithDecision("accepted"),
+		)
+	case errors.Is(err, errRateLimitInternalError) || strings.HasPrefix(err.Error(), "expected 1 response"):
+		attrs = append(attrs,
+			telemetry.WithReason(telemetry.RequestErr),
+			telemetry.WithDecision("accepted"),
+		)
+	case errors.Is(err, errTooManyRequests):
+		attrs = append(attrs,
+			telemetry.WithDecision("throttled"),
+		)
+	}
+
+	return attrs
+}
+
+func rateLimit(
+	ctx context.Context,
+	hits int,
+	rateLimit func(ctx context.Context, n int) error,
+	metadataKeys []string,
+	telemetryBuilder *metadata.TelemetryBuilder,
+) error {
+	err := rateLimit(ctx, hits)
```

Comment:
@vigneshshanmugam Following your comment: https://github.com/elastic/opentelemetry-collector-components/pull/578#discussion_r2107573655.

This is the `rateLimit` call. So the timer could start before, and stop after this call and we have the duration. This works for both local rate limiter and gubernator, instead of repeating them like we were doing before. WDYT?

Reply:
Yeah sounds good. I was maybe thinking for active/inflight reqs it might be tricky but not a deal breaker.